---
layout: post
title: "[객체지향 프로그래밍] 6주차 - 예외처리"
excerpt: "try-catch문, throw, throws, 확인/비확인예외, finally, try-with-resources"

tags:
  - [객체지향 프로그래밍, Java]

toc: true

date: 2024-02-20
last_modified_at: 2024-02-20
---
## 예외처리
### 1. 예외처리를 하는 이유
- 프로그램이 실행되는 도중에 에러가 발생하더라도,  
프로그램이 아무 이유도 모르게 갑자기 죽어버려서는 안 된다.  

  - **어떤 에러**가 **왜 발생**했는 지 알 수 있어야 한다.  

    - 따라서 <u>여러가지 에러에 프로그래머가 나름대로의 error code를 임의로 배정하고 다루는 작업</u>을 **예외처리** 라고 한다.  
    <br>
### 2. 아날로그식 예외처리
- 예외를 처리하는 전통적인 방법은,  
바로 예외가 발생하는 상황을 대비하여 따로 케이스를 나눠두는 것이다.  
예를들자면 이렇다.

```java
Scanner input = new Scanner(System.in);
int index = input.nextInt();
int[] list = {1, 2, 3};
if(index < 0 || index >= list.length) // 전통적인 예외처리
    System.out.println("잘못된 인덱스가 아닐까요?");
else 
    System.out.println(list[index]);
```

<br>

### 3. 예외를 던지고(`throw`)
- 예외를 처리하는 <u>개선된 방법</u>이 바로 **예외 던지기**이다.  

  - 일을 제대로 수행할 수 없는 경우, 프로그램이 죽는 대신에  
  `Exception`을 만들어 던지는(**`throw`**) 것을 말한다.  
    
    - `Exception`은 `java.lang` 패키지에 속한 [클래스][def]로서, 많은 서브 클래스들을 가지고 있다.  
    예를들어 `IndexOutOfBoundsException`은 `Exception`의 일종(서브클래스) 이다.  
    <br>
    
  - 예외는 다음과 같이 던질 수 있다.
  
  ```java
  // throw [Exception객체] 의 형태
  throw new IndexOutOfBoundsException();
  ```

  <br>

  - 일단 어디서든 던져진 예외는 메소드 호출 체인을 따라 위로(상위로) **전파**된다.  
  전파되어 올라가는 도중 어디서든 예외를 잡아서 처리할 수 있다.

    - 만약 예외가 던져지면, 다음 줄로 넘어가는 대신  
    **즉시** 예외처리기로 실행점이 옮겨진다.  
    (그렇다고해서, 프로그램이 즉시 종료되는 것은 아니다.)
    <br>

### 4. 예외를 잡아서 처리한다(`catch`)
- 예외가 던져졌다면, 언젠간 예외를 **잡아서 처리**해야 한다.  
예외를 처리하는 문장이 바로 **`try-catch`** 문의 **`catch`** 이다.  
예외를 잡아서 처리하는 코드를 살펴보자.

#### [1] `try-catch`

```java
try {
    // 예외가 발생 가능한 문장
    // 배열의 index 위치의 원소를 가져옴
    // IndexOutOfBoundException 이 던져질 수 있다.
    getElement(index); 
}
catch(IndexOutOfBoundsException e) {
    // 예외발생시 실행할 문장
    System.out.println("인덱스 오류");
}
```

<br>

#### [2] `try-catch-finally`

- `try-catch` 문에서 `finally` 블록이 추가 된 형태로서,  
`finally` 블록은 예외 발생 여부와 상관 없이 무조건 실행되는 특성을 가진다.

```java
try {
    // 예외가 발생 가능한 문장
}
catch(IndexOutOfBoundsException e) {
    // 예외발생시 실행할 문장
}
finally {
    // 무조건 실행할 문장. 예외 발생 여부와 관계없이 실행
}
```

- `finally` 문은 주로 `try-catch`문에서 사용한 자원을 반납하거나 뒷정리 하는데 사용된다.

<br>

#### [3] 예외에 담긴 메세지
- 예외를 던질 때, 예외 객체 생성자에 인자로서 **임의의 메세지**를 담을 수 있다.  

```java
throw new IndexOutOfBoundsException("인덱스: " + index);
```

- 그렇게 담긴 메세지는 예외처리기에서 `getMessage()` 메소드를 통해 읽을 수 있다.  
[`getMessage`][def2] 메소드는 `Exception` 클래스가 구현하고 있는 [`Throwable`][def3] 인터페이스의 메소드이다.

```java
catch(IndexOutOfBoundsException e) {
    System.out.println(e.getMessage());
}
```

- 또한 예외 객체는 **예외 상황에 대한 정보**를 가지고 있다.  
예외 상황에 대한 정보란, 예외가 발생하기 까지의 메소드 호출 체인 기록(Call Stack)을 말한다.  
  - `Throwable` 인터페이스 메소드인 [`printStackTrace()`][def4] 를 호출함으로서 정보를 출력할 수 있다.

```java
catch(IndexOutOfBoundsException e) {
    // 문자열을 반환하는 게 아닌, void 메소드이다.
    e.printStackTrace();  // Call Stack을 출력한다.
}
```

<br>

#### [4] Try-with-resources 문장

- `try`문을 실행할 때 어떠한 자원(resource)을 가지고 갈 수 있는 문장이다.  
`try(resource)`의 형태로 작성된다.  

```java
try(PrintWriter out = new PrintWriter("output.txt")) {
    out.println("~~~");
}
```

- `try` 블록을 빠져나올 때, 예외 여부에 관계없이 `out.close()` 메소드가 무조건 실행된다.  

  - `try(resource)`에서 `resource`는 `AutoClosable` 인터페이스를 구현하고 있어야 한다. (`Scanner`, `PrintWriter` 등)

  - 만약 `out.close()` 에서 예외가 던져졌을 경우에는,  
    - `try` 블록에선 예외가 발생하지 않은 경우에는  
    해당 예외가 그대로 던져진다.  
    - `try` 블록에서도 예외가 발생한 경우에는  
    `try` 블록에서 던져진 예외에 추가로 덧붙여진다.  
    참고로 덧붙여진 예외를 읽을때는 `getSuppressed()` 메소드를 호출한다.  
<br>

### 5. 예외처리를 떠넘긴다(`throws`)
- 예외를 지금 당장 처리(`catch`)하고 싶지 않을 수도 있다.
<br>

- 그래서, 예외를 처리할 때, 특정한 표시를 남김으로서  
예외를 처리하지 않고 위로 전파해버리는 방법도 존재한다.  
<br>

- 이번엔, 다른 예외가 일어나는 상황으로 예시를 들어보겠다.  
<br>
`java.util.Scanner` 클래스의 구성자 `Scanner(File f)`는 읽을 파일이 실제로 존재하지 않을 때  
`FileNotFoundException` 예외 객체를 던진다(`throw`).  
<br>

```java
public static void main(String[] args) throws FileNotFoundException {
    readFileTest();
}
static void readFileTest() throws FileNotFoundException {
    String fileName = "data.txt";
    readFile(fileName);
}
static void readFile(String fileName) throws FileNotFoundException {
    Scanner input = new Scanner(new File(fileName));
    while(in.hasNextLine())
        System.out.println(input.nextLine());
    input.close();
}
```

- 위처럼, 메소드 헤더에 자신도 **예외를 던질거라는 표시**를 단다(`throws`)  
  - 이는 `나는 이 예외를 처리하지 않고 위로 그냥 전파할 것`이란 의미의 경고문구이다.  
- 따라서 전파된 예외를 받아버린 메소드들 중 하나는 **언젠가** 예외를 잡아서(`catch`) 처리해야 한다.  
  - 만약 예외를 마지막 `main` 메소드에서 마저 `catch` 하지 않고 또다시 `throws`하여 전파하면,  
  최종적으로 시스템의 예외처리기가 알아서 예외를 잡아 처리한다.  
  <br>

### 6. 예외가 여러 개일 때
- 예외가 여러 개면 모두 나열할 수 있다.
  - (1) `throws`
  
  ```java
  public void write(String fileName) throws FileNotFoundException, EndOfFileException {
    // ~~
  }
  ```

  - (2) `catch`

  ```java
  try {
    // ~~
  }
  catch(FileNotFoundException e) {
    // ~~
  }
  catch(EndOfFileException e) {

  }
  ```

  - 단, 이 때는 순서가 문제 될 수도 있다.

    ```java
    try {
        // ~~
    }
    catch(IOException e) {
        // ~~
    }
    catch(FileNotFoundException e) {
        // ~~
    }
    ```

    - 이 코드의 경우 `IOException`이 `FileNotFoundException`의 슈퍼클래스이기 때문에,  
    첫 번재 `catch`에서 예외가 모두 걸러져  
    두 번째 `catch`에 절대로 도달할 수 없게 된다.  
    이런 경우에는 두 순서를 바꿔주어야 한다.  

- 이런 경우를 없애기 위해 예외를 통합할 수도 있다.  
  - (1) `throws`

  ```java
  public void write(String fileName) throws IOException {
    // ~~
  }
  ```

  - (2) `catch`

  ```java
  try {
    // ~~
  }
  catch(IOException e) {
    // ~~
  }
  ```  

  <br>

## 확인예외 & 비확인예외
- 모든 예외를 반드시 처리할 필요는 없다.  
만약 그렇게 된다면 처리해야할 예외가 너무 많아 코드가 길어질 것이다.  
예외에도 두 가지 종류가 있다.
<br>

- **확인예외**(Checked Exception) : 예외처리를 하도록 컴파일러가 강제하는 예외.
- **비확인예외**(Unchecked Exception) : 예외처리 여부를 컴파일러가 확인 **X**  

![exception_hierarchy][def5]  

- 앞서 알아본 예외들 중  
`IOException`은 **확인예외** 였고,  
`IndexOutOfBoundsException`은 **비확인예외** 였다.  
<br>

- **확인예외**는 반드시 적절한 예외처리를 해주어야 한다.  
그렇지 않으면 컴파일되지 않는다.
  - **확인예외**는 프로그램 오류는 아니지만,  
  *상황에 따라 이따금씩 발생*하는 오류이다. (디스크 고장, 네트워크 불안정 등)  
    - 복구가 가능하다면 예외처리기가 자동으로 복구를 시도하고,  
    복구가 불가능하다면 그냥 프로그램이 죽는게 낫다. (상세 정보 출력)  

- **비확인예외**는 예외처리기를 프로그램에 넣어 줄 필요가 없다. (해도 됨)  
  - **비확인예외**는 *프로그램 오류로 인해 발생*한다.  
    - 따라서 애써 필요하지 않은 예외처리기를 넣어주기 보다는,  
    프로그램 **오류 자체가 발생하지 않도록** 해결하는게 중요하다.  
    <br>

## Throw early, catch late
- 예외처리를 할 때의 일반적인 지침이다.  

- 되도록 예외가 일어났을 때 즉시 예외를 던지고(`throw`),  
예외를 최대한 늦게까지 전가함으로서 끌고가다가(`throws`),  
최대한 늦게 예외를 처리하는 것이(`catch`)  
좋은 프로그램이다.

<br>
<br>
<br>
<br>
<details>
<summary>주의사항</summary>
<div markdown="1">

이 포스팅은 강원대학교 정충교 교수님의 객체지향 프로그래밍 수업을 들으며 내용을 정리 한 것입니다.  
수업 내용에 대한 저작권은 교수님께 있으니,  
다른 곳으로의 무분별한 내용 복사를 자제해 주세요.
</div>
</details>

[def]: https://download.java.net/java/GA/jdk14/docs/api/java.base/java/lang/Exception.html
[def2]: https://download.java.net/java/GA/jdk14/docs/api/java.base/java/lang/Throwable.html#getMessage()
[def3]: https://download.java.net/java/GA/jdk14/docs/api/java.base/java/lang/Throwable.html#getMessage()
[def4]: https://download.java.net/java/GA/jdk14/docs/api/java.base/java/lang/Throwable.html#printStackTrace()
[def5]: https://i.imgur.com/kZK8SCF.png